cmd/kube-apiserver/app/options/validation.go

```diff
@@ -128,7 +128,7 @@ func validateTokenRequest(options *ServerRunOptions) []error {
 		errs = append(errs, errors.New("the BoundServiceAccountTokenVolume feature depends on the TokenRequest feature, but the TokenRequest features is not enabled"))
 	}
 
-	if !enableAttempted && utilfeature.DefaultFeatureGate.Enabled(features.BoundServiceAccountTokenVolume) {
+	if !enableAttempted && utilfeature.DefaultFeatureGate.Enabled(features.TokenRequest) {
```

Comment:
is this feature planned as opt-in?

`enableAttempted` appears to track if the user has passed one of the flags and if no flags are passed `enableAttempted` will be `false`. TokenRequest is already defaulted to true (beta), so unless i'm mistaken this will result in a breaking change to users that don't have the flags yet.

Reply:
yes, either BoundServiceAccoutTokenVolume results in the breaking change or TokenRequest does. making the flags required for TokenRequest will allow the tests (https://github.com/kubernetes/kubernetes/pull/95886) to pass which is using kubeadm.

FYI, these features are documented here https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#alpha-beta.

Reply:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#alpha-beta
> Set feature gate `TokenRequest=true`. (default to `true` since 1.12)

1.12 was some time ago, but did we warn the users that they should start using the flags either via a release note or a warning in the kube-apiserver? this could have made the transition smoother.

if this PR is a breaking change it would need the "action required: " prefix in the release-note.

Reply:
we had a release note in https://github.com/kubernetes/kubernetes/blob/f99ed92ed149e8cbb09dfcdcff78e6fb4024d3a0/CHANGELOG/CHANGELOG-1.12.md#other-notable-changes-15.

but didn't mention they are gonna be required in the future.

Reply:
we are going to require `--service-account-signing-key-file` after this change as most clouds have deployed `TokenRequest`?